readme.md
# ScalaCZML a CZML library in scala 

This library **ScalaCZML** reads and writes [CZML JSON](https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/CZML-Guide) entities and presents them as [Scala](http://www.scala-lang.org/) objects.

[Cesium](http://cesiumjs.org/) is a JavaScript library for creating 3D globes and 2D maps in a web browser without a plugin. 
It uses WebGL for hardware-accelerated graphics, and is cross-platform, cross-browser, 
and tuned for dynamic-data visualization. 

[CZML](https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/CZML-Guide) is a JSON schema for describing a time-dynamic graphical scene, primarily for display in a web browser running Cesium.
It describes lines, points, billboards (markers), models and other graphical primitives, and specifies how they change with time.

**ScalaCZML** makes the CZML JSON entities available as Scala objects. The library follows the specifications of reference 1. 
Typically **ScalaCZML** is for server side application development.

There is also the Cesium library in Scala, [CesiumScala](https://github.com/workingDog/CesiumScala) for client side development.

## References
 
1) [CZML guide](https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/CZML-Guide)

2) [CZML content](https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/Packet)

3) [CZML structure](https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/CZML-Structure)

## Dependencies

Using [Scala](http://www.scala-lang.org/) and the Play-JSON library of the [Play Framework](https://www.playframework.com/)

## Installation and packaging

To use the last release add the following dependency to your build.sbt:

    libraryDependencies += "com.github.workingDog" %% "scalaczml" % "0.5"

To compile and generate a jar file from the source:

    sbt package

The jar file (scalaczml_2.11-0.6-SNAPSHOT.jar) will be in the "./target/scala-2.11/" directory.

## Documentation

To generate the ScalaCZML API documentation, type "sbt doc". The documentation will be generated in 
/target/scala-2.11/api directory. **ScalaCZML** mostly follows reference 2 documentation. 

## Example

    object Example1 {
     def main(args: Array[String]) {
         // read a CZML document from a file
         val jsonDoc = Source.fromFile("/.....test4.czml").mkString
         // create a czml object from the json document
         val czml = CZML[CZMLPacket](jsonDoc)
         // create a position property
         val pos = CzmlPositions(9.3, 8.2, 7.1)
         // create a billboard property with image uri and scale fields
         val bb = new Billboard(image = "http://localhost/img.png", scale = 0.7)
         // create a czml packet
         val packet = new CZMLPacket("test packet", HashSet[CzmlProperty](pos, bb))
         // add the packet to the existing czml object
         czml.add(packet) // or czml.packets += packet
         // convert the czml object to json
         val jsczml = Json.toJson(czml)
         // write the czml as a json document to file (here to System.out)
         Util.writeCzmlToFile(czml)
         // alternatively
         //  Util.writeJsToFile(Json.prettyPrint(jsczml))
     }
    }
    
## Status

usable



CHANGES.md
Change Log
==========

## changes in 0.6-SNAPSHOT


## changes in 0.5

* ready for cross publishing for scala_2.12.1
* remove the Logger
* added convenience Integer constructors to CzmlNumber and Number
* minor update to Util.writeJsToFile
* replaced UnitQuaternionValue in Orientation to UnitQuaternion

## changes in 0.4

* added color, colorBlendMode, and colorBlendAmount properties to Model
* added ColorBlendMode
* added ShadowMode
* added shadows to Box, Corridor, Cylinder, Ellipse, Ellipsoid, Polygon, Polyline, Rectangle, Wall, Model
* added silhouetteColor and silhouetteSize to Model

## changes in 0.3

   * restructured RADIAN and DEGREES as extending GEOTYPE
   * modified Cartographic and LngLatAlt to use GEOTYPE
   * added convenience constructors for CzmlPosition and CzmlPositions using GEOTYPE, similarly in CzmlImplicits
   * added HeightReference
   * added heightReference to Billboard, Label, Model and Point

## changes in 0.2

* Breaking changes
    * renamed LineMaterial to PolylineMaterial
    * renamed ImageUri to CzmlUri
    * renamed WsenDegrees to RectangleCoordinates in czmlCore and CzmlImplicits
    * changed CZMLPacket to take ViewFrom for field viewFrom instead of CzmlCartesian
    * changed Orientation to use UnitQuaternionValue
    * changed NodeTransformation to its own Rotation type instead of Orientation
    * renamed Stripe to StripeMaterial
    * renamed Grid to GridMaterial
    * renamed SolidColor to SolidColorMaterial
    * renamed Image to ImageMaterial
    * renamed Cartesian to Cartesian3D
    * renamed PolylineArrow to PolylineArrowMaterial
    * renamed PolylineGlow to PolylineGlowMaterial
    * renamed PolylineOutline to PolylineOutlineMaterial
    * renamed Style to LabelStyle

* other changes
    * added PolylineArrow to czmlCore
    * added polylineArrow, image, grid and stripe to PolylineMaterial.
    * added implicit PolylineArrowToPolylineArrowOp to CzmlImplicits
    * added Box to czmlProperties and its BoxDimensions to czmlCore
    * added CornerType and CornerTypeValue to czmlCore
    * added Cylinder to czmlProperties
    * added Corridor to czmlProperties
    * added UnitQuaternion and UnitQuaternionValue to czmlCore
    * added Rotation to czmlCore
    * added ViewFrom to czmlProperties
    * added UriInterval to CzmlUri
    * added NearFarScalar and NearFarScalarValue to CzmlCore
    * added width, height, scaleByDistance, translucencyByDistance, pixelOffsetScaleByDistance and imageSubRegion to Billboard
    * added translucencyByDistance, pixelOffsetScaleByDistance to Label
    * added scaleByDistance, translucencyByDistance to Point
    * added maximumScale to Model
    * added BoundingRectangle and BoundingRectangleValue to CzmlCore
    * added outlineWidth to Ellipsoid
    * added RectangleCoordValue and RectangleCoordValues to czmlCore
    * added wsen to RectangleCoordinates and made wsen and wsenDegrees Option[RectangleCoordValues]
    * added RectCoordToRectCoordValOp to CzmlImplicits
    * removed ArrayOfDoubleToRect from CzmlImplicits


src/main/scala/com/kodekutters/czml/Example1.scala
package com.kodekutters.czml

import com.kodekutters.czml.czmlProperties._
import com.kodekutters.czml.CzmlImplicits._

import play.api.libs.json.Json
import scala.collection.mutable.HashSet
import scala.io.Source


/**
  * a basic example
  */
object Example1 {
  def main(args: Array[String]) {
    // read a CZML document from a file
    val jsonDoc = Source.fromFile("/.....test4.czml").mkString
    // create a czml object from the json document
    val czml = CZML[CZMLPacket](jsonDoc)
    // create a position property
    val pos = CzmlPositions(9.3, 8.2, 7.1)
    // create a billboard property with image uri and scale fields
    val bb = new Billboard(image = "http://localhost/img.png", scale = 0.7)
    // create a czml packet
    val packet = new CZMLPacket("test packet", HashSet[CzmlProperty](pos, bb))
    // add the packet to the existing czml object
    czml.add(packet) // or czml.packets += packet
    // convert the czml object to json
    val jsczml = Json.toJson(czml)
    // write the czml as a json document to file (here to System.out)
    Util.writeCzmlToFile(czml)
    // alternatively
    //  Util.writeJsToFile(Json.prettyPrint(jsczml))
  }
}